src/test/java/lambda/part3/exercise/Exercise2.java

```diff
@@ -0,0 +1,149 @@
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.junit.Assert.assertEquals;
+
+@SuppressWarnings({"unused", "WeakerAccess", "ConstantConditions"})
+public class Exercise2 {
+
+    private static class MapHelper<T> {
+
+        private final List<T> source;
+
+        private MapHelper(List<T> source) {
+            this.source = source;
+        }
+
+        public static <T> MapHelper<T> from(List<T> source) {
+            return new MapHelper<>(source);
+        }
+
+        public List<T> getMapped() {
+            return new ArrayList<>(source);
+        }
+
+        /**
+         * Создает объект для маппинга, передавая ему новый список, построенный на основе исходного.
+         * Для добавления в новый список каждый элемент преобразовывается с использованием заданной функции.
+         * ([T], (T -> R)) -> [R]
+         * @param mapping Функция преобразования элементов.
+         */
+        public <R> MapHelper<R> map(Function<T, R> mapping) {
+            //return MapHelper.from(source.stream().map(mapping).collect(Collectors.toList()));
+
+            List<R> rSource = new ArrayList<>();
+            for (T t : source) {
+                rSource.add(mapping.apply(t));
+            }
+            return MapHelper.from(rSource);
+        }
+
+        /**
+         * Создает объект для маппинга, передавая ему новый список, построенный на основе исходного.
+         * Для добавления в новый список каждый элемент преобразовывается в список с использованием заданной функции.
+         * ([T], (T -> [R])) -> [R]
+         * @param flatMapping Функция преобразования элементов.
+         */
+        public <R> MapHelper<R> flatMap(Function<T, List<R>> flatMapping) {
+            List<R> rSource = new ArrayList<>();
+            for (T t : source) {
+                rSource.addAll(flatMapping.apply(t));
+            }
+            return MapHelper.from(rSource);
+        }
+    }
+
+    @Test
+    public void mapEmployeesToLengthOfTheirFullNamesUsingMapHelper() {
+        List<Employee> employees = getEmployees();
+
+        List<Integer> lengths = MapHelper.from(employees)
+        .map(e -> e.getPerson())
+        .map(Person::getFullName)
+        .map(s -> s.length())
+        .getMapped();
+
+        assertEquals(Arrays.asList(14, 19, 14, 15, 14, 16), lengths);
+    }
+
+    @Test
+    public void mapEmployeesToCodesOfLetterTheirPositionsUsingMapHelper() {
+        List<Employee> employees = getEmployees();
+
+        List<Integer> codes = MapHelper.from(employees)
+                .flatMap(Employee::getJobHistory)
+                .map(JobHistoryEntry::getPosition)
+                .flatMap(s -> {
+                    List<Character> listChars = new ArrayList<>();
+                    for (char c : s.toCharArray()) listChars.add(c);
+                    return listChars;
+                })
```

Comment:
I'll be glad if you extract this lambda to new method and place here reference to it like ```.flatMap(Exersice2::getChars)``` - that's clearer.